### ISSUE TITLE
Pa11y fails to run on certain URLs

### ISSUE THREAD

**Issue by JordanGS** [pa11y/pa11y#166]
```
$ pa11y "https://mysite.ca/contact_add.html?contact_id=23&session_id=99999"
Welcome to Pa11y

 > PhantomJS browser created
Error: Page "https://mysite.ca/contact_add.html?contact_id=23&session_id=99999" could not be opened
    at /home/automaton/.linuxbrew/lib/node_modules/pa11y/node_modules/truffler/lib/truffler.js:149:18
    at /home/automaton/.linuxbrew/lib/node_modules/pa11y/node_modules/node-phantom-simple/node-phantom-simple.js:69:18
    at IncomingMessage.<anonymous> (/home/automaton/.linuxbrew/lib/node_modules/pa11y/node_modules/node-phantom-simple/node-phantom-simple.js:628:9)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:926:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

This doesn't look like an encoding issue to me, but it might be? I don't understand why this happens because i have other URL's that have _'s in the them as well.

**Update:**

If i do not run it with quotes "", i get the following

```
$ pa11y https://mysite.ca/contact_add.html?contact_id=23&session_id=99999
Welcome to Pa11y

 > PhantomJS browser created
 > Testing the page "https://mysite.ca/contact_add.html?contact_id=23"

Results for https://mysite.ca/contact_add.html?contact_id=23:
```

Maybe an issue with the second get parameter `session_id` because if i change it to `sessionid`, it runs fine.

**Update 2:**

It's not session_id at fault, i changed the entire thing to `https://mysite.ca/contact_add.html?contactsessionid=4232` and the error is still thrown.

@GlynnPhillips @rowanmanning , would either of you have any idea why this page cannot be opened? Thank you!

**Comment by JordanGS** [comment]
@GlynnPhillips This seems to be a problem with the authentication script we wrote before. If i use `https://mysite.ca/contact_add.html?contactsessionid=4232` as part of the pa11y command, then it get the above error? Can you think of any reason for that or how i could further debug?

**Comment by GlynnPhillips** [comment]
Hi @JordanGS

Can you confirm a few things for me please;

- Attach your config
- Is the domain you are posting the actual domain you are testing or is `mysite.ca` something you are using to hide the url you are test?
- Is the query string used to do something on the page? and if so can you confirm what it is doing?

**Comment by JordanGS** [comment]
@GlynnPhillips `mysite.ca` is a mask, not an actual url.

**config.js**
```
module.exports = {
    beforeScript: function(page, options, next) {
        var waitUntil = function(condition, retries, waitOver) {
            page.evaluate(condition, function(error, result) {
                if (result || retries < 1) {
                    waitOver();
                } else {
                    retries -= 1;
                    setTimeout(function() {
                        waitUntil(condition, retries, waitOver);
                    }, 200);
                }
            });
        };

        page.evaluate(function() {
            var user = document.querySelector('#inputUsername');
            var password = document.querySelector('#inputPassword');
            var submit = document.querySelector('#Login');

            user.value = 'rvolpe';
            password.value = 'Test1234';

            submit.click();
        }, function() {
            waitUntil(function() {
                // Wait until the login has been success and the /news.html has loaded
                return window.location.href === 'https://mysite.ca/news.html';
            }, 20, function() {
                // Redirect to the page contact.html page
                page.evaluate(function() {
                    window.location.href = 'https://mysite.ca/contact_add.html?contactsessionid=4232';
                });
                waitUntil(function() {
                    // Wait until the contact.html has been loaded before running pa11y
                    return window.location.href === 'https://mysite.ca/contact_add.html?contactsessionid=4232';
                }, 20, next);
            });
        });
    }
}
```

`pally --config config.js 'https://mysite.ca/contact_add.html?contactsessionid=4232'`

throws the error

however, `pally --config config.js 'https://mysite.ca/news.html'`

seems to work. I have it being kicked off by jenkins now for 100 runs for the entire app to verify if that is the cause.

i wouldn't have thought that the original url `'https://mysite.ca/contact_add.html?contactsessionid=4232'` would cause the problem though.

**Comment by GlynnPhillips** [comment]
Ok thanks @JordanGS. Would it be possible to give me the actual url that's being tested if we did that privately? 

You could join the pa11y slack (http://pa11y.slack.com) and private message it to me there if your ok with that.

**Comment by JordanGS** [comment]
@GlynnPhillips I asked my manager if i could do that and no go, there is no way to give someone that is outside the organization, everything is hosted on intranet. I also cannot create a user for someone outside the organization and it would be a breach of privacy to give you the real credentials and vpn access of an actual user. Sorry :(

I have a temporary solution as mentioned above and I'll keep working on it on my end and updating as i find more information. It seems to be a problem with the original url passed to pa11y. It has to be the login page or the landing page. If it is any other page on the website, even if there is a redirect to the login page, it will break for this instance.

**Comment by mfairchild365** [comment]
So, I just had a similar issue and figured out that it was due to some SSL configuration. This is how I fixed it. I created a config.json file and passed it with command line via the --config option:

```
"phantom": {
    "parameters": {
        "ssl-protocol": "any"
    }
}
```

The default appears to be `tlsv1`.

**Comment by joseluisbolos** [comment]
Hi @JordanGS - Is this still an issue?

**Comment by JordanGS** [comment]
@joseluisbolos Don't know, i've been busy with security testing and such that i haven't been using pa11y since i couldn't get it to do what i want at the moment. I'll have some free time in the coming weeks that i'll update then.

**Comment by joseluisbolos** [comment]
Excellent, no worries. If you ever get round to have a look at this again, please make sure you're using the latest version of phantomjs (2.1.1 at the time of writing). There were some `ssl-protocol`-related changes on phantom 1.9.8 so I'm wondering if that could have anything to do with it.

There's currently an open PR in pa11y ( #190 ) to include `phantomjs-prebuilt`. This will ensure that we're always using the latest version, but this hasn't been merged yet.

Also, it would be good if you could have a look at [mfairchild365's comment](https://github.com/pa11y/pa11y/issues/166#issuecomment-261069394) and try all of the following phantom configs and see what results do you get with each one of them:

```
"ssl-protocol": "any"
"ssl-protocol": "tlsv1"
"ssl-protocol": "ssl3"
```

I'll be closing this issue for now. Feel free to reopen it if you keep having problems.

**Comment by joseluisbolos** [comment]
For future reference of anyone finding this issue, we've just released a version [4.1.0](https://github.com/pa11y/pa11y/blob/master/CHANGELOG.md#410-2016-11-21) of pa11y that:
* Uses `phantomjs-prebuilt` to install phantomjs, so it's not necessary to install it separately anymore, and we'll always be sure that we're using the latest version.
* Reverted #62, which could be causing problems like the one in this issue #166 on servers with tlsv1.0 disabled.
